### PR TITLE
[FLINK-3226] Casting support for arithmetic operators

### PR DESCRIPTION
**Pull request by twalthr**
This PR implements auto-casting for numeric arithmetic operators.

@fhueske @vasia mergable?

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/OperatorCodeGen.scala [hunk 1]
```diff
@@ -18,20 +18,56 @@
...
+    }
+    // Numeric arithmetic
+    else if (isNumeric(left) && isNumeric(right)) {
+      val leftType = left.resultType.asInstanceOf[NumericTypeInfo[_]]
+      val rightType = right.resultType.asInstanceOf[NumericTypeInfo[_]]
+
+      generateOperatorIfNotNull(nullCheck, resultType, left, right) {
+      (leftTerm, rightTerm) =>
+        // insert auto casting for "narrowing primitive conversions"
+        if (leftType != rightType) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/OperatorCodeGen.scala]: Shouldn't we compare and cast to the result type?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/OperatorCodeGen.scala [hunk 2]
```diff
@@ -18,20 +18,56 @@
...
+          // leftType can not be casted to rightType automatically -> narrow
+          if (!leftType.shouldAutocastTo(rightType)) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/OperatorCodeGen.scala]: Wouldn't this downcast `double` to `short` given an expression: `double + short`?

Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/OperatorCodeGen.scala]: Just checked, double is actually downcasted.
The following test fails:
```
val t = env.fromElements((100000.0d: Double, 1: Byte)).toTable.select('_1 + '_2)
val expected = "100001.0"
```

**Comment by twalthr** [comment]
@fhueske you were totally right. I reworked the casting again.

**Comment by fhueske** [comment]
Thanks for the update. LGTM, merging this

**Comment by fhueske** [comment]
Merged, please close @twalthr

**Pull request closed without merging**